File: #1658.minimum-operations-to-reduce-x-to-zero.cpp
class Solution {
public:
    int minOperations(vector<int>& nums, int x) {

        int size = nums.size();
        int sum = std::accumulate(begin(nums), end(nums), 0);
        int left = 0;
        int right = 0;
        int result = INT_MAX;
        while (left <= right) {
            if (sum >= x) {
                if (sum == x) {
                    result = std::min(result, left + size - right);
                }

                if (right < size) {
                    sum -= nums[right++];
                } else {
                    break;
                }
            } else {
                sum += nums[left++];
            }
        }
        return result == INT_MAX ? -1 : result;
    }
};

File: #1658.minimum-operations-to-reduce-x-to-zero.js
/**
 * @param {number[]} nums
 * @param {number} x
 * @return {number}
 */
var minOperations = function(nums, x) {
    let length = nums.length;
    let index = 0;
    let complement_length = -1;
    let complement_sum = 0;
    let total = nums.reduce((left, right) => left + right, 0);
    for (let right = 0; right < length; right++ ) {
        complement_sum += nums[right];
        while (index <= right && complement_sum > total-x) {
            complement_sum -= nums[index++];
        }

        if (complement_sum == total - x) {
            complement_length = Math.max(complement_length,
                right - index + 1);
        }
    }
    return complement_length == -1 ? -1 :
        length - complement_length;
};
